src/emulator/registry.ts

```diff
@@ -16,9 +17,11 @@ export class EmulatorRegistry {
       throw new FirebaseError(`Emulator ${instance.getName()} is already running!`, {});
     }
 
+    // Start the emulator and wait for it to grab its assigned port.
     await instance.start();
-    this.set(instance.getName(), instance);
+    await controller.waitForPortClosed(instance.getInfo().port);
```

Comment:
I also moved the `waitForPortClosed()` bit into the registry so that we only mark an emulator as started once it takes its port.